Hidden friends (#204)
hsf-training/cpluspluscourse: talk/objectorientation/operators.tex

  \end{block}
\end{frame}

\begin{frame}[fragile]
  \frametitlecpp[98]{Friend declarations}
  \begin{block}{Concept}
    \begin{itemize}
      \item Functions/classes can be declared \mintinline{cpp}{friend} within a class scope

>       \item They gain access to all private members

Reviewer: ```suggestion
      \item They gain access to all private/protected members
```

      \item Useful for operators such as \mintinline{cpp}{a + b}

>       \item Don't abuse this! Don't modify an object in a friend!

Reviewer: > Don't modify an object in a friend!

What is the rational for this? Why shouldn't I grant a different class the ability to modify me?

Author: Of course you can. And there's cases where it might make sense. But more often than not, I see the friends being abused as a convenient way out of a proper refactoring. Especially in HEP in older codes.
What I want to say is:
- They are great for operators where you mostly need read access.
- If you use write access, all bets are off. Don't do stupid things / unexpected things / thread-unsafe things / ...

I didn't know how to say this in one sentence.

Reviewer: Hmm, I see. Then maybe you can add:
```suggestion
      \item Don't abuse this! Avoid modifying an object in a friend, use the objects member functions instead.
```
Maybe also split the two sentences into two bullet points.

Author: They now read:
```diff
+      \item Don't abuse friends to go around a wrongly designed interface
+      \item Avoid unexpected modifications of class state in a friend
```

    \end{itemize}
  \end{block}
  \begin{block}{operator+ as a friend}

>     \begin{cppcode*}{gobble=2,linenos=false}

Reviewer: I would prefer line numbers :)

>       struct Complex {
>         float m_real; float m_imag;
>         friend Complex operator+(Complex & a, Complex & b);
>       };
> 
>       Complex operator+(Complex & a, Complex & b) {
>         return Complex(a.m_real+b.m_real,a.m_imag+b.m_imag);
>       }

Reviewer: ```suggestion
      struct Complex {
        float m_real; float m_imag;
        friend Complex operator+(Complex & a, Complex & b);
      };

      Complex operator+(Complex & a, Complex & b) {
        return Complex(a.m_real+b.m_real,a.m_imag+b.m_imag);
      }
```
I think you miss `const` on the parameters.
And the point of friendship would be stronger, if `m_real` and `m_imag` were private.

Author: The const doesn't fit on the slide, so I removed it. Also, at this point, we haven't explained constness yet. OK to leave it out or shrink the text?

Author: I'll check it private fits ...

Reviewer: C++ is so interdependent, I think we cannot topsort all the content to only use features that have been previously explained. And I would like to avoid having code on the slides that people should not write in practice, like an operator with a mutable argument. So yeah, I would prefer shrinking the slide to not having `const` :)

Author: Manage to fit everything.

  \end{block}
\end{frame}

\begin{frame}[fragile]
  \frametitlecpp[98]{Friend declarations}
  \begin{block}{Concept}
    \begin{itemize}
      \item Functions/classes can be declared \mintinline{cpp}{friend} within a class scope
      \item They gain access to all private/protected members
      \item Useful for operators such as \mintinline{cpp}{a + b}
      \item Don't abuse friends to go around a wrongly designed interface
      \item Avoid unexpected modifications of class state in a friend
    \end{itemize}
  \end{block}

>   \begin{block}{operator+ as a friend}

Reviewer: On the other slide, you used a different font:
```suggestion
  \begin{block}{\texttt{operator+} as a friend}
```

Author: That's because your suggestion to use `operator<<` was typeset as `operator¡¡`. I unified it now.

    \small
    \begin{cppcode*}{gobble=2}
      class Complex {

>         float m_real, m_imag;

Reviewer 2: This does not compile if you do not declare a constructor
```cpp
Complex( float, float );
```

Reviewer 2: Fixed in #229

Reviewer: @hageboeck originally had ` // + Complex(float, float) ...` on that line as a comment, and I suggested to make `Complex` a `struct`. We must have missed this in the end.

>         float m_real, m_imag; public: Complex(float, float);

Reviewer: Since you are not constructing a `Complex` on this slide, you could omit the constructor if you want:
```suggestion
        float m_real, m_imag;
```
